### PR TITLE
Merge of XPlat master and TTD Node branches.

### PR DESCRIPTION
**Pull request by mrkmarron**
Merge in changes needed for time-travel debugging on xplat.

### CODE REVIEW

**Comment by mrkmarron** [comment]
@agarwal-sandeep, @digitalinfinity, @ianwjhalliday please take a look.

File: src/node_zlib.cc [hunk 1]
```diff
@@ -181,10 +181,16 @@ class ZCtx : public AsyncWrap {
       // sync version
       ctx->env()->PrintSyncTrace();
       Process(work_req);
+#if ENABLE_TTD_NODE
```
Comment (agarwal-sandeep) [inline src/node_zlib.cc]: No need for ENABLE_TTD_NODE as the macro TTD_NATIVE_BUFFER_ACCESS_NOTIFY is noop if ENABLE_TTD_NODE is not defined

File: src/node_buffer.h [hunk 2]
```diff
@@ -19,9 +19,25 @@ NODE_EXTERN bool HasInstance(v8::Local<v8::Value> val);
 NODE_EXTERN bool HasInstance(v8::Local<v8::Object> val);
 NODE_EXTERN char* Data(v8::Local<v8::Value> val);
 NODE_EXTERN char* Data(v8::Local<v8::Object> val);
+
+#if ENABLE_TTD_NODE
+NODE_EXTERN void TTDAsyncModRegister(v8::Local<v8::Object> val,
+    byte* initialModPosition);
```
Comment (agarwal-sandeep) [inline src/node_buffer.h]: > byte* initialModPosition); [](start = 3, length = 27)

Function parameters on new line should be on column after opening braces, see other APIs. Please fix all places.

File: src/node_buffer.h [hunk 3]
```diff
@@ -19,9 +19,25 @@ NODE_EXTERN bool HasInstance(v8::Local<v8::Value> val);
...
+
 NODE_EXTERN size_t Length(v8::Local<v8::Value> val);
 NODE_EXTERN size_t Length(v8::Local<v8::Object> val);
 
+#if ENABLE_TTD_NODE
+// Notify us that a native buffer access (which we don't currently understand/support) happened.
+#define TTD_NATIVE_BUFFER_ACCESS_NOTIFY(X) JsTTDCheckAndAssertIfTTDRunning("Unsupported raw buffer access -- investigate this!!!\n")
+#else
+#define TTD_NATIVE_BUFFER_ACCESS_NOTIFY(X)
+#endif
```
Comment (agarwal-sandeep) [inline src/node_buffer.h:33]: Why not combine this with previous if, keeps code under one ifdef?

File: src/node_buffer.h [hunk 4]
```diff
@@ -19,9 +19,25 @@ NODE_EXTERN bool HasInstance(v8::Local<v8::Value> val);
...
+NODE_EXTERN void TTDSyncDataModNotify(v8::Local<v8::Object> val,
+    UINT32 index, UINT32 count);
+#endif
+
 NODE_EXTERN size_t Length(v8::Local<v8::Value> val);
 NODE_EXTERN size_t Length(v8::Local<v8::Object> val);
 
+#if ENABLE_TTD_NODE
+// Notify us that a native buffer access (which we don't currently understand/support) happened.
+#define TTD_NATIVE_BUFFER_ACCESS_NOTIFY(X) JsTTDCheckAndAssertIfTTDRunning("Unsupported raw buffer access -- investigate this!!!\n")
```
Comment (agarwal-sandeep) [inline src/node_buffer.h:30]: Lines are > 80 char, please run cpplint on all changed .cc and .h files and eslint on .js files.

**Comment by kunalspathak** [comment]
`ENABLE_TTD_NODE` should be set in `chakrashim.gyp` only for chakracore. node-v8 should be able to build with your changes.

**Comment by boingoing** [comment]
LGTM

**Pull request merged**